### PR TITLE
Add Generalised Linear Models to FSharpStats

### PR DESCRIPTION
**Pull request by LibraChris**
**Changes introduced in this PR**

- Implemented QR-based solver for GLMs
- Added multiple tests for various GLM link and variance functions
- Added functions for GLM statistics and documentation

**Description**

This PR introduces the GLM (Generalized Linear Models) module in FSharp.Stats. The key features include the implementation of an Iteratively Reweighted Least Squares (IRLS) solver on a QR-based solver for GLMs, enhancing the accuracy and performance of statistical modeling. Extensive tests have been added and updated to ensure robustness, covering various link functions and variance calculations. Documentation has been thoroughly updated to assist users in leveraging these new features effectively.

**[Required]** please make sure you checked that
 - [X] The project builds without problems on your machine

**[Optional]**
 - [X] Added unit tests regarding the added features

---

### CODE REVIEW

**Comment by LibraChris** [comment]
Hey @bvenn, would you mind checking this out? Please squish all the commits if possible to avoid confusion. If there is anything I should add or change please let me know

File: src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs
Comment (kMutagene) [inline src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs]: Is this the correct location for a function like this? Are these functions so specialized that nobody ever needs to use them in another context? if not, i'd suggest putting them in the managed LinearAlgebra module, otherwise you can also set the whole QR module here to internal

Comment (LibraChris) [reply src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs]: While some of the functions in QR are highly specialised, some, qrAlternative and solveLinearQR, aren't. The module is needed for future development when I will add a solver based on singular value decomposition, but at the moment it is redundant. 

I will move the the specified functions to LinearAlgebra but keep the rest there if that´s ok.

Comment (kMutagene) [reply src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs]: Allrighty. You can then mark the whole specialized module as internal to avoid leaking any of that stuff in the public API.

Comment (LibraChris) [reply src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs]: Doing that would also remove it from the testing I on the stepwise function. You can have a look and tell me if it is ok now.

Comment (kMutagene) [reply src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs]: You can expose internals to other projects in the repo by adding this attribute in the project file:

```xml
<ItemGroup> 
  <InternalsVisibleTo Include="TestProject1" /> 
</ItemGroup>
```

Comment (LibraChris) [reply src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs]: @kMutagene hey, I changed the code accordingly. Could you have another look if its better now?

File: src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs
Comment (kMutagene) [inline src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs:687]: In general for XML docs, use the provided tags for annotating things such as input arguments an return values, see more [here](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/xml-documentation#recommended-tags)

**Review by kMutagene — CHANGES_REQUESTED** [review]

File: src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs
Comment (LibraChris) [reply src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs:687]: Like this?

File: tests/FSharp.Stats.Tests/GeneralisedLinearModels.fs
Comment (kMutagene) [inline tests/FSharp.Stats.Tests/GeneralisedLinearModels.fs:146]: I do not understand this test setup. I think you should base the first conditional on the expected value and not on the actual value, e.g. "If the expected value is infinity then actual should be infinity", same for the tests below

**Review by kMutagene — APPROVED** [review]
LGTM from a maintenance standpoint - i cannot assess the implementation itself or the Linear Algebra though, so as long as your tests pass i think this is fine, test setup is a little weird but maybe i just do not understand the specifics

**Comment by LibraChris** [comment]
Hey @bvenn, anything you would change before the merge?

**Comment by bvenn** [comment]
> Hey @bvenn, anything you would change before the merge?

Sorry, I was busy during the last weeks. I'll check out your PR this week.

File: src/FSharp.Stats/Algebra/LinearAlgebra.fs
Comment (bvenn) [inline src/FSharp.Stats/Algebra/LinearAlgebra.fs:268]: I'm not sure if I get whats happening here:

1. qLengths is a zero-vector that gets filled iteratively for each `i` in line 262
2. in line 256 setqOfA is called with the current i. However in l 247 qLength is called at position `i` which always should be 0, because it was initialized with 0. Just after the determination of qLength the value at this specific index is replaced in line 262.

Obviously I miss a step, but is there a possibility to create the qLength by Vector.init (...) rather than keeping it mutable and accessing/mutating it at multiple positions? If not, I'm happy to merge it, but I got confused during my attempt to understand whats happening here..

Comment (bvenn) [reply src/FSharp.Stats/Algebra/LinearAlgebra.fs:268]: Could you additionally set into <remarks>-brackets a reference to this implementation and a short description what the benefit, drawback is when using this implementation over others.

File: src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs
Comment (bvenn) [inline src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs]: Is this field actually called PersonOfZ? I did no research but it seems to me as if standardError and ZScore would be associated to a p-value?

File: src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs
Comment (bvenn) [inline src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs:59]: I'm not sure if this is supported by [XML-Documentation](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/xml-documentation#recommended-tags). In VSCommunity, the styling is not applied. 

I checked if the API-Reference was displayed correctly using `dotnet fsdocs watch --eval` and it didn't work. 

Does it work on your machine? If not, please use the allowed tags only

File: src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs
Comment (bvenn) [inline src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs:208]: I assume the `1./2.` and later `-1.-1.` are left on purpose to make the implementation easier to understand :+1:

File: src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs
Comment (bvenn) [inline src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs:264]: You could use the build in `sign` function

File: src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs
Comment (bvenn) [inline src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs:371]: `e` isn't used, is it correct?

**Review by bvenn — CHANGES_REQUESTED** [review]
GLM + documentation looks good, but I have a general thought: Why is the GLM module assigned to the Fitting-namespace rather than Testing? I know that GLMs are directly linked to regression analysis, but I personally would search at Testing when looking for GLMs. But maybe I'm wrong here, lets discuss.

**Comment by bvenn** [comment]
When watching the docs `dotnet fsdocs watch --eval` I noticed the API-Reference is broken. Nothing is displayed. I'm not sure if the cause lies within this PR or some error happened earlier. If you find time, you can check if it worked prior to your commits.

**Comment by bvenn** [comment]
> When watching the docs `dotnet fsdocs watch --eval` I noticed the API-Reference is broken. Nothing is displayed. I'm not sure if the cause lies within this PR or some error happened earlier. If you find time, you can check if it worked prior to your commits.

The issue seems to be older than this PR or it is due to my setup. Could you check the docs in your machine?

**Comment by kMutagene** [comment]
> When watching the docs `dotnet fsdocs watch --eval` I noticed the API-Reference is broken. Nothing is displayed. I'm not sure if the cause lies within this PR or some error happened earlier. If you find time, you can check if it worked prior to your commits.

> > The issue seems to be older than this PR or it is due to my setup. Could you check the docs in your machine?

Try using the build target (`build.cmd watchdocs`), which should make sure that the necessary binaries are built before creating the docs

**Comment by bvenn** [comment]
I still get the following error during `build watchdocs`:

"System.Xml.XmlException: The ',' character, hexadecimal value 0x2C, cannot be included in a name. Line 15278, position 99"
I think I fixed an issue similar to this within the last months.

@LibraChris your branch is 42 commits behind, if you could rebase, the issue may be gone

File: src/FSharp.Stats/Algebra/LinearAlgebra.fs
Comment (LibraChris) [reply src/FSharp.Stats/Algebra/LinearAlgebra.fs:268]: I do not have a direct reference since I had a look at the mathematical explanation for QR via Gram Schmidt and tried implementing it myself. The only real upside using this method instead of the in FSharp.Stats established one is the difference in the output Dimensions in R. An actual reference would be the numpy implantation of the reduced qr https://numpy.org/doc/stable/reference/generated/numpy.linalg.qr.html .

I also think that this implemtation needs to be optimised in the future to perfect the GLM.

Comment (LibraChris) [reply src/FSharp.Stats/Algebra/LinearAlgebra.fs:268]: > I'm not sure if I get whats happening here:
> 
> 1. qLengths is a zero-vector that gets filled iteratively for each `i` in line 262
> 2. in line 256 setqOfA is called with the current i. However in l 247 qLength is called at position `i` which always should be 0, because it was initialized with 0. Just after the determination of qLength the value at this specific index is replaced in line 262.
> 
> Obviously I miss a step, but is there a possibility to create the qLength by Vector.init (...) rather than keeping it mutable and accessing/mutating it at multiple positions? If not, I'm happy to merge it, but I got confused during my attempt to understand whats happening here..

The mutation was done to simplify the function to match the mathematical explanation better. This is one of the things that could be updated when I get to it.

File: src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs
Comment (LibraChris) [reply src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs]: You are right, this is supposed to be Pearson. I fixed this.

File: tests/FSharp.Stats.Tests/FSharp.Stats.Tests.fsproj
Comment (bvenn) [inline tests/FSharp.Stats.Tests/FSharp.Stats.Tests.fsproj]: Could you please fix these indentations

File: src/FSharp.Stats/Algebra/LinearAlgebra.fs
Comment (bvenn) [inline src/FSharp.Stats/Algebra/LinearAlgebra.fs]: Thanks for the updates!
I'm going to merge this PR if you just could add proper XML docs like seen in:
https://github.com/fslaborg/FSharp.Stats/blob/c5444cc52c2e58b92d34f10c0c6de90fc4625433/src/FSharp.Stats/Fitting/LinearRegression.fs#L128-L145

The same for line 222 f.

**Review by bvenn — CHANGES_REQUESTED** [review]

File: src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs
Comment (bvenn) [reply src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs:264]: ping

File: src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs
Comment (bvenn) [reply src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs:371]: ping

File: src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs
Comment (bvenn) [reply src/FSharp.Stats/Fitting/GeneralisedLinearModel.fs:687]: yes 👍

**Review by bvenn — APPROVED** [review]

**Pull request merged**
